### DISCUSSION TITLE
How do I mix --select-1 and become(vim{})?

### DISCUSSION THREAD

**Discussion by elazar125** [junegunn/fzf#3803] [Q&A] [answered]
I'm trying to use `fzf --select-1 --bind 'enter:become(nvim {+})'` to fuzzy-find files, while skipping over the fzf bit if I know what I'm looking for (and can type something sufficiently unique). However, when I trigger the `--select-1` with a unique query, I never get the chance to press enter to trigger the become(), and the file name is just printed to stdout instead of being fed into vim.

Is there an event I can bind to that'll let me become(vim) with only one result?

**Comment by LangLangBart** [comment]
> Is there an event I can bind to that'll let me become(vim) with only one result?


```sh
man fzf | less --pattern "AVAILABLE EVENTS"
```

You could try the `one`[^1] event.

```sh
echo -n "$USER" | fzf --bind 'one:become:rev <<<{}'
# airap
```
The `load`[^2] event combined with a `transform`[^3] action and an exported environment variable[^4] by `fzf` might work as well.

```sh
echo -n 'fzf' | fzf --bind 'load:transform:[[ $FZF_TOTAL_COUNT -eq 1 ]] && echo "become:man {}"'
```

All of the above methods have the minor downside of briefly flickering the UI.

[^1]: [changelog 0.39.0](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0390)
[^2]: [changelog 0.36.0](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0360)
[^3]: [changelog 0.45.0](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0450)
[^4]: [changelog 0.46.0](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0460)

**Reply by elazar125** [reply]
Thanks for the answer!

Unfortunately the `load`/`transform` version doesn't seem to do it, I'm still just getting the path printed to stdout.

The `one` option technically does work, but it feels a little odd to use. With `one` filtering down to one option also triggers the `become()`, without taking an explicit action to select the result. I foresee this being an issue if I typo the filter as I'm looking for my file. When trying to call my function with a query pre-typed, I still have to press enter to run the function, so (assuming I'm paying attention) I have a chance to catch typos. I'll try it out for a bit in case it just takes some getting used to!

Really I'm hoping there's some trick I'm missing to bind explicitly to the `--select-1` behaviour without also affecting the way I would use the TUI, but maybe the answer is adding a bit of bash, and not just built-in fzf behaviour.

**Reply by LangLangBart** [reply]
> Unfortunately the `load`/`transform` version doesn't seem to do it, I'm still just getting the path printed to stdout.

Can u share the command ?

> With `one` filtering down to one option also triggers the `become()`

That is an important point, you could `unbind` the `one` event with another event or hotkey, e.g. `--bind 'change:unbind:one'`

```bash
echo "foo\nbar" | fzf  --bind 'change:unbind:one' --bind 'one:become:echo {}'
```

**Reply by elazar125** [reply]
The full command I used for `load`/`transform` is
```
find_file_and_edit() {
    fzf \
        --multi \
        --select-1 \
        --query "$*" \
        --bind 'enter:become("$EDITOR" {+}),load:transform:[[ $FZF_TOTAL_COUNT -eq 1 ]] && echo "become:nvim {}"' \
        --preview 'bat --color=always {}'
}
```
Also, splitting it into two `--bind`s didn't work either.

Although playing around with it more, `one` also isn't doing it (I must have not tested calling the function with a query earlier). The `one` binding is only triggering when I filter within fzf, not when the `--select-1` triggers.

**Reply by elazar125** [reply]
In case this helps at all:
```
$ fzf --version
0.52.0 (bcda25a5)
```

**Reply by junegunn** [reply, marked as answer]
`one` and `unbind` strategy works just fine.

```sh
find_file_and_edit() {
  fzf \
      --sync \
      --multi \
      --query "$*" \
      --bind 'change:unbind(one),enter:become("$EDITOR" {+}),one:become("$EDITOR" {})' \
      --preview 'bat --color=always {}'
}
```

---

Your command has some errors.

* You should remove `--select-1` since you're relying on the event binding mechanism within the finder
* You should use `result` event instead of `load` event because you're interested in the "result" for your initial `--query`.
    * Loading and matching runs asynchronously in parallel, so `load` event doesn't guarantee that you have the result for your query
* Use `FZF_MATCH_COUNT` instead of `FZF_TOTAL_COUNT` to see the number of matches for your initial `--query`
* `{}` needs escaping because the path can contain double-quotes
* You still need to "unbind" `result` event.
* In case your source program takes a while to populate the list, use `--sync` option. e.g. `(sleep 1; ls) | find_file_and_edit`

```sh
find_file_and_edit() {
  fzf \
      --sync \
      --multi \
      --query "$*" \
      --bind 'change:unbind(result),enter:become("$EDITOR" {+}),result:transform:[[ $FZF_MATCH_COUNT -eq 1 ]] && echo "become:\"\$EDITOR\" \{}"' \
      --preview 'bat --color=always {}'
}
```

Same thing, but worse.

**Reply by elazar125** [reply]
Looks like removing `--select-1` was the missing ingredient, using only `one` with `change:unbind:one` is working now. Thank you!
